File: Data Structures & Algorithms/binary-tree-diameter/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    diameterOfBinaryTree(root) {
        let res = 0
        
        const depth = (node) => {
            if(node === null)
                return 0
            
            // tính chiều sâu hai nhánh
            let left = depth(node.left)
            let right = depth(node.right)

            //Cập nhật đường kính
            res = Math.max(res, left+right)

            //return chiều sau của node
            return 1+Math.max(left,right)
        }

        depth(root)
        return res
        
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public int GoodNodes(TreeNode root) {
        return DFS( root, root.val);
    }

    public int DFS(TreeNode node, int max)
    {
        if(node == null)
            return 0;
        int count = 0;

        if(node.val >= max)
        {
            count = 1;
            max = node.val;
        }

        count += DFS(node.left, max);
        count += DFS(node.right, max);

        return count; 
    }
}
